lib/ipam/ipam.go

```diff
@@ -1524,3 +1526,66 @@ func decideHostname(host string) (string, error) {
 	log.Debugf("Using hostname=%s", hostname)
 	return hostname, nil
 }
+
+// GetUtilization returns IP utilization info for the specified pools, or for all pools.
+func (c ipamClient) GetUtilization(ctx context.Context, args GetUtilizationArgs) ([]*PoolUtilization, error) {
+	var pools []*PoolUtilization
+
+	// Read all pools.
+	allPools, err := c.pools.GetAllPools()
+	if err != nil {
+		log.WithError(err).Errorf("Error getting IP pools")
+		return nil, err
+	}
+
+	// Identify the ones we want and create a PoolUtilization for each of those.
+	wantAllPools := len(args.Pools) == 0
+	wantedPools := set.FromArray(args.Pools)
+	for _, pool := range allPools {
+		if wantAllPools ||
+			wantedPools.Contains(pool.Name) ||
+			wantedPools.Contains(pool.Spec.CIDR) {
+			pools = append(pools, &PoolUtilization{
+				Name: pool.Name,
+				CIDR: net.MustParseNetwork(pool.Spec.CIDR).IPNet,
+			})
+		}
+	}
+
+	// If we've been asked for all pools, also report utilization for any allocation
+	// blocks for which there is no longer an IP pool.  Note: following code depends
+	// on this being at the end of the list; otherwise it will suck in allocation
+	// blocks that should be reported under other pools.
+	if wantAllPools {
+		pools = append(pools, &PoolUtilization{
+			Name: "orphaned allocation blocks",
+			CIDR: net.MustParseNetwork("0.0.0.0/0").IPNet,
+		})
+	}
+
+	// Read all allocation blocks.
+	blocks, err := c.client.List(ctx, model.BlockListOptions{}, "")
+	if err != nil {
+		return nil, err
+	}
+	for _, kvp := range blocks.KVPairs {
+		b := kvp.Value.(*model.AllocationBlock)
+		log.Debugf("Got block: %v", b)
+
+		// Find which pool this block belongs to.
+		for _, pool := range pools {
+			if b.CIDR.IsNetOverlap(pool.CIDR) {
+				log.Debugf("Block CIDR %v belongs to pool %v", b.CIDR, pool.Name)
+				pool.Blocks = append(pool.Blocks, BlockUtilization{
+					CIDR:      b.CIDR.IPNet,
+					Capacity:  b.NumAddresses(),
+					Available: len(b.Unallocated),
+				})
+				break
+			} else {
+				log.Debugf("Block CIDR %v does not overlap pool CIDR %v", b.CIDR, pool.CIDR)
```

Comment:
I suggest you remove this `log.Debugf` cause the `log.Debugf` above is good enough.